File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/block/BlockContinueTest.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.readChain;

class BlockContinueTest {

    @Test
    public void blockContinue() {
        TokenChain chain = readChain("continue;");

        Context context = new Context(chain.get());
        BlockWhile blockWhile = getBlockWhile(context);
        BlockContinue blockContinue = new BlockContinue(blockWhile, chain.get(), null);
        blockContinue.read();

        assertErrors(context);
    }

    @Test
    public void blockContinueWithoutLoop_Fail() {
        TokenChain chain = readChain("continue;");

        Context context = new Context(chain.get());
        BlockContinue blockContinue = new BlockContinue(context, chain.get(), null);
        blockContinue.read();

        assertErrors(context, Error.continueOutOfPlace);
    }

    @Test
    public void blockContinueUnexpectedToken_Fail() {
        TokenChain chain = readChain("continue hello;");

        Context context = new Context(chain.get());
        BlockWhile blockWhile = getBlockWhile(context);
        BlockContinue blockContinue = new BlockContinue(blockWhile, chain.get(), null);
        blockContinue.read();

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void blockContinueUnexpectedEndOfTokens_Fail() {
        TokenChain chain = readChain("continue");

        Context context = new Context(chain.get());
        BlockWhile blockWhile = getBlockWhile(context);
        BlockContinue blockContinue = new BlockContinue(blockWhile, chain.get(), null);
        blockContinue.read();

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    private BlockWhile getBlockWhile(Context context) {
        TokenChain chain = readChain("while(true);");

        return new BlockWhile(context, chain.get(), null);

    }
}

File: SimplexApp/simplex/parser/logic/block/BlockContinue.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const Context = require("simplex/parser/logic/Context.js");
const ContextSupport = require("simplex/support/ContextSupport.js");
const Error = require("simplex/parser/logic/error/Error.js");
const BlockContinue = require("simplex/parser/logic/block/BlockContinue.js");
const BlockWhile = require("simplex/parser/logic/block/BlockWhile.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;

test('blockContinue', () => {
    let chain = readChain('continue;');
    let context = new Context(chain.get());
    let blockWhile = getBlockWhile(context);
    let blockContinue = new BlockContinue(blockWhile, chain.get(), null);
    blockContinue.read();
    assertErrors(context);
});
test('blockContinueWithoutLoop_Fail', () => {
    let chain = readChain('continue;');
    let context = new Context(chain.get());
    let blockContinue = new BlockContinue(context, chain.get(), null);
    blockContinue.read();
    assertErrors(context, Error.continueOutOfPlace);
});
test('blockContinueUnexpectedToken_Fail', () => {
    let chain = readChain('continue hello;');
    let context = new Context(chain.get());
    let blockWhile = getBlockWhile(context);
    let blockContinue = new BlockContinue(blockWhile, chain.get(), null);
    blockContinue.read();
    assertErrors(context, Error.unexpectedToken);
});
test('blockContinueUnexpectedEndOfTokens_Fail', () => {
    let chain = readChain('continue');
    let context = new Context(chain.get());
    let blockWhile = getBlockWhile(context);
    let blockContinue = new BlockContinue(blockWhile, chain.get(), null);
    blockContinue.read();
    assertErrors(context, Error.unexpectedEndOfTokens);
});
function getBlockWhile(context) {
    let chain = readChain('while(true);');
    return new BlockWhile(context, chain.get(), null);
}